sn_client/src/files/upload.rs

```diff
@@ -186,210 +218,170 @@ impl FilesUpload {
+                }
+            }
 
-            let payee = if let Some(payee) = payee_map
-                .iter()
-                .find(|itr| itr.0 == chunk_info.name)
-                .map(|result| result.1)
+            let task_result = if let Some(result) = progress_tasks(
+                &mut get_store_cost_receiver,
+                &mut paid_chunk_receiver,
+                &mut upload_chunk_receiver,
+            )
+            .await
             {
-                payee
+                result
             } else {
-                error!(
-                    "Cannot find payee of {:?} among the payee_map",
-                    chunk_info.name
-                );
-                continue;
+                return Err(ClientError::FailedToReadFromNotificationChannel);
             };
 
-            // Spawn a task for each chunk to be uploaded
-            let handle = tokio::spawn(Self::upload_chunk(
-                files_api,
-                chunk_info,
-                payee,
-                verify_store,
-            ));
-
-            self.progress_uploading_chunks(false).await?;
-
-            self.uploading_chunks.push(handle);
-        }
-
-        Ok(())
-    }
-
-    /// Progresses the uploading of chunks. If the number of ongoing uploading chunks is less than the batch size,
-    /// it pays for the next batch and continues. If an error occurs during the upload, it will be returned.
-    ///
-    /// If `drain_all` is true, will wait for all ongoing uploads to complete before returning.
-    async fn progress_uploading_chunks(&mut self, drain_all: bool) -> Result<()> {
-        while drain_all || self.uploading_chunks.len() >= self.batch_size {
-            if let Some(result) = self.uploading_chunks.next().await {
-                // bail if we've had any errors so far
-                match result? {
-                    (chunk_info, Ok(())) => {
+            match task_result {
+                TaskResult::GetStoreCostOK((chunk_info, cost)) => {
+                    let _ = on_going_get_cost.remove(&chunk_info.name);
+                    trace!(
+                        "Upload task got chunk {:?}'s store_cost {:?}",
+                        chunk_info.name,
+                        cost.2
+                    );
+                    if cost.2.cost != NanoTokens::zero() {
+                        pending_to_pay.push((chunk_info, cost));
+                    } else if for_failed_chunks {
                         self.send_event(FileUploadEvent::Uploaded(ChunkAddress::new(
                             chunk_info.name,
-                        )))
-                        .await?;
+                        )));
+                        uploaded += 1;
+                    } else {
+                        self.send_event(FileUploadEvent::AlreadyExistsInNetwork(
+                            ChunkAddress::new(chunk_info.name),
+                        ));
+                        skipped += 1;
                     }
```

Comment:
Can we add comments regarding this special case if the cost is zero & based on if it is a re-upload. We can just copy over the previous comment.